### PR TITLE
ec2_instance - add support for `placement` and `license_specifications`

### LINKED ISSUE #1824: Support For Additional Placement Options and LicenseSpecifications
### Summary

In order to launch mac instances using the `amazon.aws.ec2_instance` module, support for HostGroupResourceArn placement option and LicensConfigurations would be required as an alternative to only specifying tenancy as dedicated to offer more control on instance launch.

### Issue Type

Feature Idea

### Component Name

ec2_instance

### Additional Information

```yaml
- name: start a mac instance with a host resource group and license configuration
  amazon.aws.ec2_instance:
    name: "public-compute-instance"
    key_name: "prod-ssh-key"
    vpc_subnet_id: subnet-5ca1ab1e
    instance_type: mac1.metal
    security_group: default
    placement:
      hostresourcegrouparn: arn:aws:resource-groups:us-east-1:123456789012:group/MyResourceGroup
    license_specifications:
      - licence_configuration_arn: 
    image_id: ami-123456
    tags:
      Environment: Testing
```

### Code of Conduct

- [X] I agree to follow the Ansible Code of Conduct

### PR DESCRIPTION
**Pull request by mfortin**
##### SUMMARY

Adding placement options and license specifications support

Fixes #1824 

##### ISSUE TYPE

- Feature Pull Request

##### COMPONENT NAME

ec2_instance

##### ADDITIONAL INFORMATION

```paste below

```

### CODE REVIEW

File: plugins/modules/ec2_instance.py [hunk 1]
```diff
@@ -2117,6 +2201,18 @@ def main():
...
+        placement=dict(
+            type="dict",
+            options=dict(
+                affinity=dict(type="str"),
+                availability_zone=dict(type="str"),
+                group_name=dict(type="str"),
+                host_id=dict(type="str"),
+                host_resource_attr=dict(type="str"),
+                partition_number=dict(type="int"),
+                tenancy=dict(type="str", choices=["dedicated", "default"]),
```
Comment (tremble) [inline plugins/modules/ec2_instance.py:2213]: This puts tenancy in two separate places, could this be taken from the tenancy setting above?

**Review by tremble — CHANGES_REQUESTED** [review]
Thanks for taking the time to submit this PR,

If possible, please could you add some integration tests for the placement options?

The tests are just Ansible roles living in: 
tests/integration/targets

There's already a number of ec2_instance_... tests in there.

File: plugins/modules/ec2_instance.py (hunk 1 above)
Comment (mfortin) [reply plugins/modules/ec2_instance.py:2213]: It does, I didn't want to introduce a breaking change, unless it is ok?
The AWS SDK has the `tenancy` attribute inside the `placement` attributes, I wanted to be as close as possible to the AWS python SDK.

Comment (tremble) [reply plugins/modules/ec2_instance.py:2213]: We can deprecate the other option, use this if set, use the other if not, throw an error if both are set.

**Comment by mfortin** [comment]
@tremble @gravesm Is there a way to mock resources for the integration tests? There are no modules right now for PlacementGroups, ResourceGroups nor LicenseManager, which makes testing a challenge.

**Review by hakbailey — CHANGES_REQUESTED** [review]
See a couple of inline comments, but one larger request: instead of adding new modules here, can we use the existing ec2_placement_group and ec2_placement_group_info modules in the [community.aws collection](https://github.com/ansible-collections/community.aws)?

File: plugins/modules/ec2_instance.py [hunk 2]
```diff
@@ -258,6 +258,7 @@
   tenancy:
     description:
       - What type of tenancy to allow an instance to use. Default is shared tenancy. Dedicated tenancy will incur additional charges.
+      - This field is deprecated, use `placement` instead.
```
Comment (mfortin) [inline plugins/modules/ec2_instance.py:261]: @hakbailey I added a deprecation message to `tenancy`

File: plugins/modules/ec2_instance.py [hunk 3]
```diff
@@ -325,7 +326,52 @@
   placement_group:
     description:
       - The placement group that needs to be assigned to the instance.
+      - This field is deprecated, use `placement` instead.
```
Comment (mfortin) [inline plugins/modules/ec2_instance.py:329]: @hakbailey also added deprecation message to `placement_group`

File: plugins/modules/ec2_instance.py [hunk 4]
```diff
@@ -2115,8 +2190,18 @@ def main():
                 threads_per_core=dict(type="int", choices=[1, 2], required=True),
             ),
         ),
-        tenancy=dict(type="str", choices=["dedicated", "default"]),
```
[the remark this answers is not in the document]
Comment (mfortin) [reply plugins/modules/ec2_instance.py:2118]: done

File: plugins/modules/ec2_instance.py [hunk 5]
```diff
@@ -2115,8 +2190,18 @@ def main():
...
-        placement_group=dict(type="str"),
```
[the remark this answers is not in the document]
Comment (mfortin) [reply plugins/modules/ec2_instance.py:2119]: done

File: plugins/modules/ec2_instance.py [hunk 6]
```diff
@@ -2146,6 +2278,8 @@ def main():
             ["image_id", "image"],
             ["exact_count", "count"],
             ["exact_count", "instance_ids"],
+            ["tenancy", "placement"],
+            ["placement_group", "placement"],
```
Comment (mfortin) [inline plugins/modules/ec2_instance.py:2284]: @hakbailey Added the checks to validate only one of tenancy/placement_group or placement is defined

File: tests/integration/targets/ec2_instance_license_specifications/tasks/main.yml [hunk 7]
```diff
@@ -0,0 +1,29 @@
+- module_defaults:
```
[the remark this answers is not in the document]
Comment (mfortin) [reply tests/integration/targets/ec2_instance_license_specifications/tasks/main.yml:1]: Added validation against the expected error

**Comment by gravesm** [comment]
I believe the permissions problem is because of the placement group. Looking into getting this fixed.

**Comment by jillr** [comment]
recheck

**Comment by mfortin** [comment]
@jillr it worked! Thanks a bunch!

File: plugins/modules/ec2_instance.py (hunk 1 above)
Comment (mfortin) [reply plugins/modules/ec2_instance.py:2251]: done 👍

**Review by mfortin — COMMENTED** [review]
I think this is ready to ship!

File: plugins/modules/ec2_instance.py (hunk 2 above)
Comment (tremble) [inline plugins/modules/ec2_instance.py:261]:
```suggestion
      - This field is deprecated, use I(placement) instead.
```

File: plugins/modules/ec2_instance.py (hunk 3 above)
Comment (tremble) [inline plugins/modules/ec2_instance.py:329]:
```suggestion
      - This field is deprecated, use I(placement) instead.
```

**Review by tremble — APPROVED** [review]

File: changelogs/fragments/1825-ec2_instances.yml [hunk 8]
```diff
@@ -0,0 +1,3 @@
+minor_changes:
+  - ec2_instance - add support for additional `placement` options and `license_specifications` in run instance spec [#1824](https://github.com/ansible-collections/amazon.aws/issues/1824).
+  - ec2_instance - deprecation of `tenancy` and `placement_group` in favor of `placement` attribute.
```
Comment (tremble) [inline changelogs/fragments/1825-ec2_instances.yml:3]:
```suggestion
  - ec2_instance - add support for additional ``placement`` options and ``license_specifications`` in run instance spec (https://github.com/ansible-collections/amazon.aws/issues/1824).
  - ec2_instance - deprecation of ``tenancy`` and ``placement_group`` in favor of ``placement`` attribute  (https://github.com/ansible-collections/amazon.aws/pull/1825)..
```

File: plugins/modules/ec2_instance.py [hunk 9]
```diff
@@ -325,7 +326,52 @@
   placement_group:
     description:
       - The placement group that needs to be assigned to the instance.
+      - This field is deprecated and will be removed in a release after 2025-12-01, use I(placement) instead.
     type: str
+  placement:
```
Comment (alinabuzachis) [inline plugins/modules/ec2_instance.py:331]: Can you please add `version_added: 7.0.0`?

**Review by alinabuzachis — APPROVED** [review]

**Pull request merged**